### PR TITLE
Cstr - add proof from_bytes_with_null

### PR DESCRIPTION
**Pull request by rajathkotyal**
Towards #150 

Verifies that the CStr safety invariant holds after calling :

from_bytes_with_nul --> core::ffi::c_str module

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: library/core/src/ffi/c_str.rs [hunk 1]
```diff
@@ -882,6 +882,34 @@ mod verify {
...
+        let mut buffer: [u8; MAX_SIZE] = kani::any();
+
+        // Iterate over all possible slice lengths from 1 to MAX_SIZE
+        for slice_length in 1..=MAX_SIZE {
+            // Ensure the slice is valid by setting the last byte to null
+            buffer[slice_length - 1] = 0;
+
+            // Assume all bytes before the last are non-null
+            for i in 0..slice_length - 1 {
+                kani::assume(buffer[i] != 0);
```
Comment (tautschnig) [inline library/core/src/ffi/c_str.rs:899]: Won't this assumption evaluate to false on the second iteration? After all, you just set byte 0 to `0` in the first iteration.

Comment (rajathkotyal) [reply library/core/src/ffi/c_str.rs]: Ah. I forgot to re initialize.

However, when I did reinitialize, I got the `CBMC appears to have run out of memory.` error. So I had to reduce the number MAX_SIZE to 16. and also simplify the function a bit. Have updated the code, pl review.

File: library/core/src/ffi/c_str.rs [hunk 2]
```diff
@@ -882,7 +882,32 @@ mod verify {
...
     }
-  
+
+    // pub const fn from_bytes_with_nul(bytes: &[u8]) -> Result<&Self, FromBytesWithNulError>
+    #[kani::proof]
+    #[kani::unwind(17)]  
+    fn check_from_bytes_with_nul() {
+        const MAX_SIZE: usize = 16;
+        let mut string: [u8; MAX_SIZE] = kani::any();
+        let null_pos: usize = kani::any_where(|x| *x < MAX_SIZE);
```
Comment (tautschnig) [inline library/core/src/ffi/c_str.rs:892]: Since you aren't doing the loop anymore, can you go back to 32?

Comment (rajathkotyal) [reply library/core/src/ffi/c_str.rs:892]: I tried doing that, but it fails due to insufficient memory with and without loops(Im running on a 32GB Ram M2 machine), Hence limited it to 16

Comment (rajathkotyal) [reply library/core/src/ffi/c_str.rs:892]: I tried doing that, but it fails due to insufficient memory, with and without loops(Im running on a 32GB Ram M2 machine), Hence limited it to 16.

I believe this is due to the use of memchar in the original function, which could expand to larger state spaces with more bits, and also the function calls the `from_bytes_with_nul_unchecked` function internally. which contains a loop that iterates from the end of the slice backwards to ensure no interior null bytes exist.

**Review by tautschnig — APPROVED** [review]

File: library/core/src/ffi/c_str.rs (hunk 1 above)
Comment (rajathkotyal) [reply library/core/src/ffi/c_str.rs]: Yes, but figured explicit checking makes it more aligned with the function.

File: library/core/src/ffi/c_str.rs [hunk 3]
```diff
@@ -885,7 +885,32 @@ mod verify {
             assert!(c_str.is_safe());
         }
     }
-  
+
+    // pub const fn from_bytes_with_nul(bytes: &[u8]) -> Result<&Self, FromBytesWithNulError>
+    #[kani::proof]
+    #[kani::unwind(17)]  
+    fn check_from_bytes_with_nul() {
```
Comment (celinval) [inline library/core/src/ffi/c_str.rs:892]: Is it possible to extend this harness to any combination of bytes up to a certain size?

File: library/core/src/ffi/c_str.rs [hunk 4]
```diff
@@ -885,7 +885,32 @@ mod verify {
...
+        const MAX_SIZE: usize = 16;
+        let mut string: [u8; MAX_SIZE] = kani::any();
+        let null_pos: usize = kani::any_where(|x| *x < MAX_SIZE);
```
Comment (celinval) [inline library/core/src/ffi/c_str.rs:895]: you could use `any_slice` instead.

File: library/core/src/ffi/c_str.rs [hunk 5]
```diff
@@ -885,7 +885,32 @@ mod verify {
...
+        
+        string[null_pos] = 0;
+        // check to make sure all bytes are non null
+        for i in 0..null_pos {
+            string[i] = kani::any_where(|x| *x != 0);
+        }
+        
+        let slice = &string[..=null_pos];
+        let result = CStr::from_bytes_with_nul(slice);
+        assert!(result.is_ok());
```
Comment (celinval) [inline library/core/src/ffi/c_str.rs:905]: The functional verification is not necessary, but it is nice to have. If you want to keep it, I would suggest that you do the following, match the result, if it is `Ok`, check that the input has a null byte in the position corresponding to the length of the new cstr. If it is an `Err`, you can double check that all bytes are non-null.

**Review by celinval — COMMENTED** [review]
Thanks!

File: library/core/src/ffi/c_str.rs (hunk 4 above)
Comment (rajathkotyal) [reply library/core/src/ffi/c_str.rs:895]: sure, that was our initial approach as well, something like : 
```
    #[kani::proof]
    #[kani::unwind(17)] 
    fn check_from_bytes_with_nul2() {
        const MAX_SIZE: usize = 16;
        let string: [u8; MAX_SIZE] = kani::any();
        let slice = kani::slice::any_slice_of_array(&string);

        let result = CStr::from_bytes_with_nul(slice);
        if let Ok(c_str) = result {
            assert!(c_str.is_safe());
        }
    }
```

After some tests I think this above code should suffice, I tried doing extra validation by validating non-null within the string, but it seems to not make a difference.

File: library/core/src/ffi/c_str.rs (hunk 3 above)
Comment (rajathkotyal) [reply library/core/src/ffi/c_str.rs:892]: Sure, will do. Also note that it fails for bit sizes >= 32

Comment (celinval) [reply library/core/src/ffi/c_str.rs:892]: Why does it fail? Which error are you getting?

**Review by celinval — APPROVED** [review]

**Review by tautschnig — APPROVED** [review]

**Pull request merged**